### ISSUE TITLE
Enhancing the New Developer Experience

### ISSUE THREAD

**Issue by ccarterlandis** [chaoss/augur#229]
This is a continuation of issue #176.

To start, here a few areas where I think we could make the new developer experience better:
- Auto-generate an `augur.config.json` based on the provided plugins
- A more comprehensive overview of the frontend and backend architecture (what goes where and why)
- Harden installation scripts and the `Makefile` to avoid issues like #180

**Comment by Nebrethar** [comment]
I've been trolling around for installation issues (at least on Ubuntu!). #232 #233 

Will be focusing on documentation over the weekend.

It seems like documentation is most needed at the moment, and deployment a close second. Opinions, other things to focus on from anyone?

**Comment by ccarterlandis** [comment]
After a good discussion earlier today with @howderek and @sgoggins, we've got some ideas about ways to make the new developer experience better.

1). We've decided not to get rid of the Anaconda integration, as we feel supporting Anaconda is preferable to supporting 4 sets of Python installation instructions (macOS, Ubunto, centOS, and Debian) and we would rather have Anaconda handle all of that. However, as part of this decision, we're going to add a few steps in the docs to help new developers check that Anaconda is correctly installed on their machine. These will be in the form of "Attempt to run `xyz` Anaconda command to ensure you've gotten far enough; then, try `abc` Anaconda command to ensure..." so on and so forth. 

2). We're going to move the installation pieces of the `Makefile` into their own separate files in an attempt to more clearly break down what's going on when installing which will allow us to more easily pinpoint installation issues.

3). I can't read @sgoggins's handwriting here. I think it says "Test" something. 😉 

![2 8_notes](https://user-images.githubusercontent.com/31522931/52516327-94ca2f80-2bee-11e9-87d0-48d8421e1eff.jpg)



4). Update the documentation! We'll have some fresh install docs to write, and when I write those up I'll go through the rest of them and do some more housekeeping.

**Comment by ccarterlandis** [comment]
I've added a preliminary version of item 1 above to the installation section of the Python documentation, under a "A Note about Anaconda".

**Comment by TeemuKoivisto** [comment]
Hi, thought I could to input my own experience installing Augur and this PR looked like a place to put it.

Simply, I was unable to run Augur locally, something something about missing facade database. No explanations whatsoever how to install it, the website for it in the docs is down https://opendata.missouri.edu/ and in general I found the installation instructions to be quite hard to follow.

Disclaimer: I did not use Vagrant since I don't want to use it. I'd rather use Docker and Docker Compose. Sadly though the `docker-compose.yml` was full of variables I had no idea what to fill with. I have a local GHTorrent MariaDB running, but where exactly I write its configuration. GH_TORRENT_PLUS? And apparently I should create another database and user called augur locally?

There was also some spelling errors in the docs and even an incorrect SQL statement: `GRANT ALL PRIVILEDGES ON augur.* TO 'augur'@'localhost';` (no D here).

Not to sound too harsh or anything, but this was my short summary of hardships I endured. If it was possible to streamline the instructions with clear-cut steps for local installation I'd very much enjoy it. =)

The reason I'm doing this in the first place is for researching different types of open-source project analysis software/stuff out there. I am a part of a group doing a school project in which we're trying to come up with some models to analyze open-source projects, we've done some stuff here https://github.com/datascienceproject2019-codescoop/codescoop-models but yeah, still WIP 😄

**Comment by ccarterlandis** [comment]
Hi @TeemuKoivisto! Your feedback is very much appreciated, we are always on the lookout for problems and pain points in our installation process. 

Full disclosure: I'm not sure when the Docker installation instructions were last updated. We prefer new users to use Vagrant, but of course that's not everyone's preference! The Vagrant stuff is relatively new and I've been focusing most of my attention on that; after the Vagrant installation process has been hardened, I'll focus on updating the Docker installation instructions. 

As far as Facade goes, that's not a project we maintain or contribute to, we just host a version of it ourselves. Lines 11-30 of the Vagrant provisioning file located [here](https://github.com/chaoss/augur/blob/hotfix/util/packaging/vagrant/provision.sh) have updated commands for installing lightweight versions of both Facade and GHTorrent, which might be useful.  I'm not sure why that link is down, I will look into it.

If you don't mind me asking, which parts specifically were hard to follow? I will target those first.

Again, thanks for the feedback!

**Comment by TeemuKoivisto** [comment]
Hey and @ccarterlandis thank you for quick response, always nice to encounter actively maintained project =].

Hmm well my initial reaction to the README was that it was complicated. I like numbered lists, like a lot. Also I try to keep my own READMEs minimum in words since I know the person installing the system isn't probably interested in my chatter, even though how relevant it might seem. Sadly. So I'd shorten the Vagrant installation instructions into 1-2-3-etc steps, with a table of contents. I only recently found that at least VisualCode supports auto-generation of TOCs in markdown files, which is pretty cool.

So as a quick example:
1. Git clone and install Vagrant configuration:
```
git clone https://github.com/chaoss/augur.git
cd augur
make vagrant
```
2. Wait for the Vagrant installation finish. After it does, you'll be automatically logged into the VM. Navigate to the working directory which is synced between the VM and your local environment and start the server:
```
sudo su -
cd /vagrant/augur
augur
```
3. This will generate a default configuration file called `augur.config.json`. Use `sample.config.json` to fill in the missing configuration. You'll need at least x and y. Follow the instructions below to install the MariaDB database with GHTorrent data. If you're interested in doing developing your own *and so on*

And then perhaps move the more explanatory details to other block say "Explanation of the installation process". Also then I'd add a short installation guide for local development without Vagrant with environment specific details left out. The current local installation guide could be the "detailed version" which was linked from the README. And it should use the same 1-2-3 steps with installation kept as simple as possible to not to overwhelm the silly programmer's brains with too much noise :).

Also this one made me feel quite uneasy:
```make install-dev-admin # some libraries require a root install.```

If something requires sudo I'd very much prefer knowing what and why and could that be avoided somehow. Not just running some magic script that prompts for sudo privileges. Especially when I see sudo being used for installing Node.js which last time I checked wasn't required for Ubuntu (can't recall was it because I used nvm or why).

But the biggest problem was the missing instructions for the configuration variables: which are required and how the databases were set up. It would help if there even was some placeholder values in the docker-compose.yml or something. Also running the app on localhost:80 is a bit unorthodox in dev setup.

**Comment by ccarterlandis** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Update installation instructions](https://trello.com/c/pVbd3Zxl/65-update-installation-instructions)

**Comment by TeemuKoivisto** [comment]
I am unable to view that card but anyway, by restoring the facade database I was able to get the augur to run. Which is nice but we are probably not going to do network or code/commit analysis for our project. Maybe something akin to libraries.io. But thanks!

**Comment by ccarterlandis** [comment]
Whoops, didn't see it left that comment! I'm glad you were able to get it to run. If it's alright with you, after I update the docs sometime this week I'd like your input as to where we can continue to enhance the process.

**Comment by TeemuKoivisto** [comment]
Sure, I can take a look. =)

**Comment by sgoggins** [comment]
How are we on this issue? I think we have a few things to do yet, but I am feeling like it's time we close this issue with the vagrant fix in pull request #270 .

**Comment by ccarterlandis** [comment]
There are still a few documentation updates I'd like to make before I think this can be marked as done.

**Comment by ccarterlandis** [comment]
We've undergone a huge re-architecting since this issue was opened, and as a part of this process we've made numerous improvements to the NDE. While we will continue to focus on the enhancing NDE moving forward, the data in this issue is now significantly out of date.
